### PR TITLE
Fix race condition in service_manager.

### PR DESCRIPTION
**Pull request by jacob314**

### CODE REVIEW

**Comment by jacob314** [comment]
Fixes 
https://github.com/flutter/devtools/issues/2500
Also fixes recent test failures for inspector_service_test and other tests that exposed the timing issue.
As general cleanup also removed cases where streams were used where ValueNotifier was a better abstraction.

File: packages/devtools_app/lib/src/service_manager.dart [hunk 1]
```diff
@@ -564,24 +563,38 @@ class IsolateManager {
...
 
-class ServiceExtensionManager {
+class ServiceExtensionManager extends Disposer {
   VmServiceWrapper _service;
   IsolateManager _isolateManager;
 
-  bool _firstFrameEventReceived = false;
+  // Id so it is easy to avoid long running requests that were started
+  // for a previous vm service connection.
+  int vmServiceId = 0;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/service_manager.dart:587]: Maybe I'm missing something, but I don't see where this is ever modified?

File: packages/devtools_app/lib/src/service_manager.dart [hunk 2]
```diff
@@ -564,24 +563,38 @@ class IsolateManager {
...
-  final Map<String, StreamController<bool>> _serviceExtensionController = {};
-  final Map<String, StreamController<ServiceExtensionState>>
-      _serviceExtensionStateController = {};
+  bool _checkForFirstFrameStarted = false;
+  bool get _firstFrameEventReceived => _firstFrameReceived.isCompleted;
+  Completer<void> _firstFrameReceived = Completer<void>();
+  Future<void> get firstFrameReceived => _firstFrameReceived.future;
 
-  final Map<String, ValueListenable<bool>> _serviceExtensionListenables = {};
+  final Map<String, ValueNotifier<bool>> _serviceExtensionAvailable = {};
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/service_manager.dart:594]: nit: not code you wrote, but put types on the right hand side, here and elsewhere

File: packages/devtools_app/lib/src/service_manager.dart [hunk 3]
```diff
@@ -735,31 +706,55 @@ class ServiceExtensionManager {
...
+    _hasServiceExtensionNotifier(name).value = true;
 
     if (_enabledServiceExtensions.containsKey(name)) {
       // Restore any previously enabled states by calling their service
       // extension. This will restore extension states on the device after a hot
       // restart. [_enabledServiceExtensions] will be empty on page refresh or
       // initial start.
-      return _callServiceExtension(name, _enabledServiceExtensions[name].value);
+      return await _callServiceExtension(
+          name, _enabledServiceExtensions[name].value);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/service_manager.dart:753]: nit: trailing comma

File: packages/devtools_app/lib/src/service_manager.dart [hunk 4]
```diff
@@ -918,9 +914,9 @@ class ServiceExtensionManager {
       await _callServiceExtension(name, value);
     }
 
-    final StreamController<ServiceExtensionState> streamController =
-        _getServiceExtensionStateController(name);
-    streamController.add(ServiceExtensionState(enabled, value));
+    // TODO(jacobr): we might want to be a little smarter about when the value
+    // hasn't really changed.
+    _serviceExtensionState(name).value = ServiceExtensionState(enabled, value);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/service_manager.dart:919]: I believe that ValueNotifier is smart about this and only notifies listeners if the value is changed. As long as ServiceExtensionState implements equality, this should be taken care of automatically.

File: packages/devtools_app/lib/src/service_manager.dart [hunk 5]
```diff
@@ -936,95 +932,45 @@ class ServiceExtensionManager {
   }
 
   ValueListenable<bool> hasServiceExtensionListener(String name) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/service_manager.dart:934]: can we rename to hasServiceExtensionListenable?

File: packages/devtools_app/test/support/mocks.dart [hunk 6]
```diff
@@ -437,11 +435,10 @@ class FakeServiceExtensionManager extends Fake
     implements ServiceExtensionManager {
   bool _firstFrameEventReceived = false;
 
-  final Map<String, StreamController<bool>> _serviceExtensionController = {};
-  final Map<String, StreamController<ServiceExtensionState>>
+  final Map<String, ValueNotifier<ServiceExtensionState>>
       _serviceExtensionStateController = {};
 
-  final Map<String, ValueListenable<bool>> _serviceExtensionListenables = {};
+  final Map<String, ValueNotifier<bool>> _serviceExtensionAvailable = {};
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/support/mocks.dart:441]: nit: types on RHS

File: packages/devtools_app/lib/src/service_manager.dart [hunk 7]
```diff
@@ -936,95 +932,45 @@ class ServiceExtensionManager {
...
-    );
+  void vmServiceOpened(VmServiceWrapper service, ConnectedApp connectedApp) {
+    _checkForFirstFrameStarted = false;
+    cancel();
+    _connectedApp = connectedApp;
+    _service = service;
+    autoDispose(service.onExtensionEvent.listen(_handleExtensionEvent));
+    addAutoDisposeListener(
+        hasServiceExtensionListener(extensions.didSendFirstFrameEvent),
+        _maybeCheckForFirstFlutterFrame);
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/service_manager.dart:970]: done

File: packages/devtools_app/lib/src/service_manager.dart (hunk 1 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/service_manager.dart:587]: nope that code was incomplete. Cleaned it up and switched to tracking the main isolate directly in the service manager as the list of service extensions is only valid for the main isolate anyway and some of the logic was just crazy if you did change isolates.

File: packages/devtools_app/lib/src/service_manager.dart (hunk 2 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/service_manager.dart:594]: done

File: packages/devtools_app/lib/src/service_manager.dart (hunk 3 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/service_manager.dart:753]: done

File: packages/devtools_app/lib/src/service_manager.dart (hunk 4 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/service_manager.dart:919]: Added equals and hashcode to ServiceExtensionState and removed todo.

File: packages/devtools_app/lib/src/service_manager.dart (hunk 5 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/service_manager.dart:934]: done. renamed to hasServiceExtension

File: packages/devtools_app/test/support/mocks.dart (hunk 6 above)
Comment (jacob314) [reply packages/devtools_app/test/support/mocks.dart:447]: done.

**Comment by jacob314** [comment]
ptal

**Review by kenzieschmoll — APPROVED** [review]

File: packages/devtools_testing/lib/service_manager_test.dart [hunk 8]
```diff
@@ -201,17 +201,15 @@ Future<void> runServiceManagerTests(FlutterTestEnvironment env) async {
       await env.tearDownEnvironment();
     });
 
-    // TODO(kenz): once hot restart tests are fixed, add a hot restart test
-    // that verifies the state of service extensions after a hot restart.
-    // TODO(jacobr): uncomment out the hotRestart tests once
+    // TODO(kenz): add a hot restart test that verifies the state of service
+    // extensions after a hot restart.
     // https://github.com/flutter/devtools/issues/337 is fixed.
```
Comment (kenzieschmoll) [inline packages/devtools_testing/lib/service_manager_test.dart:206]: looks like something happened here with the comment? the bug references a todo line that was deleted

**Review by kenzieschmoll — APPROVED** [review]

File: packages/devtools_testing/lib/service_manager_test.dart (hunk 8 above)
Comment (jacob314) [reply packages/devtools_testing/lib/service_manager_test.dart:206]: cleaned up.

**Pull request merged**